sdcm/scan_operation_thread.py

```diff
@@ -1,163 +1,316 @@
-                                   self.number_of_rows_read, self.time_elapsed)
-                    operation_event.message = f"{type(self).__name__} operation ended successfully"
+                    self.fetch_result_pages(result=result, read_pages=self.fullscan_stats.read_pages)
+                    scan_op_event.message = f"{type(self).__name__} operation ended successfully"
                 except Exception as exc:  # pylint: disable=broad-except
+                    op_stat.exceptions.append(exc)
                     msg = str(exc)
-                    msg = f"{msg} while running Nemesis: {db_node.running_nemesis}" if db_node.running_nemesis else msg
-                    operation_event.message = msg
+                    msg = f"{msg} while running " \
+                          f"Nemesis: {self.db_node.running_nemesis}" if self.db_node.running_nemesis else msg
+                    scan_op_event.message = msg
 
-                    if db_node.running_nemesis or any(s in msg.lower() for s in ERROR_SUBSTRINGS):
-                        operation_event.severity = Severity.WARNING
+                    if self.db_node.running_nemesis or any(s in msg.lower() for s in ERROR_SUBSTRINGS):
+                        scan_op_event.severity = Severity.WARNING
                     else:
-                        operation_event.severity = Severity.ERROR
-
-    def run(self):
-        end_time = time.time() + self.duration
-        while time.time() < end_time and not self.termination_event.is_set():
-            self.db_node = random.choice(self.db_cluster.nodes)
-            self.read_pages = random.choice([100, 1000, 0])
-            self.run_scan_operation()
-            self.log.debug('Executed %s number: %s', self.scan_event.__name__, self.scans_counter)
-            time.sleep(self.interval)
-
-    def start(self):
-        self._thread.start()
-
-    def join(self, timeout=None):
-        return self._thread.join(timeout)
+                        scan_op_event.severity = Severity.ERROR
+                finally:
+                    duration = time.time() - start_time
+                    self.fullscan_stats.time_elapsed += duration
+                    self.fullscan_stats.scans_counter += 1
+                    op_stat.nemesis_at_end = self.db_node.running_nemesis
+                    op_stat.duration = duration
+                    op_stat.success = not bool(op_stat.exceptions)  # success is True if there were no exceptions
+                    self.update_stats(op_stat)
+                    return op_stat  # pylint: disable=lost-exception
+
+    def update_stats(self, new_stat):
+        self.fullscan_stats.stats.append(new_stat)
+
+    def run_scan_operation(self, cmd: str = None) -> FullScanOperationStat:
+        return self.run_scan_event(cmd=cmd or self.randomly_form_cql_statement(), scan_event=self.scan_event)
 
+    def fetch_result_pages(self, result, read_pages):
+        self.log.debug('Will fetch up to %s result pages.."', read_pages)
```

Comment:
nitpick: there is an unnecessary `"`

Reply:
Fixed

Reply:
i still see it:
```suggestion
        self.log.debug('Will fetch up to %s result pages..', read_pages)
```

Reply:
Ah sorry, didn't push the change. Should be fixed now.